### PR TITLE
feat(rules): add tally/ruby/state-paths-not-writable-as-non-root

### PR DESCRIPTION
**Pull request by tinovyatkin**
## Summary

Adds [`tally/ruby/state-paths-not-writable-as-non-root`](https://tally.wharflab.com/rules/tally/ruby/state-paths-not-writable-as-non-root/) — Section 4.2 of [design-docs/43-ruby-on-docker.md](../blob/main/design-docs/43-ruby-on-docker.md).

A Rails app stage that switches to a non-root `USER` but `COPY`s application content without `--chown` leaves the Rails state directories (`tmp`, `log`, `storage`, `db`) root-owned. Rails crashes on first request when ActionView, ActiveStorage, ActiveRecord, or `bin/rails db:prepare` tries to write to them. The Rails generator template handles this via `COPY --chown=rails:rails ... /rails /rails`; forks that drop the flag ship runtime-broken images.

- **Compliance:** `--chown=<user>...` on the COPY (with bare-user form `--chown=rails` accepted), or a runtime `chown -R <user> tmp log storage db` covering all four canonical state directories.
- **Suppressions:** stages without an explicit USER, root USER, non-Ruby stages, non-Rails-shaped images (no `WORKDIR /rails`/`/app`, no rails-style ENTRYPOINT, no `RAILS_ENV`).
- **Auto-fix:** `FixSuggestion`. Inserts `--chown=<user>:<user>` after the `COPY` keyword. Marked suggestion (not safe) because user/group names may need adjustment for unusual setups.

## Test plan

- [x] `make build`
- [x] `make test` — ruby package and integration suite pass
- [x] `make lint` — 0 issues
- [x] Lint and fix fixtures: `internal/integration/fixtures/{lint,fix}/ruby-state-paths-not-writable-as-non-root/`

## Release Notes

* **New Features**
  * Added a linting rule to detect Rails/Ruby containers that switch to non-root users without ensuring critical state directories remain writable after copying application code.

* **Documentation**
  * Added documentation for the new rule with Dockerfile examples, fix suggestions, and guidance on proper ownership configurations.

[Review Change Stack](https://app.coderabbit.ai/change-stack/wharflab/tally/pull/675)

### CODE REVIEW

**Comment by tinovyatkin** [comment]
@coderabbitai review

**Comment by tinovyatkin** [comment]
@codex review

File: internal/rules/tally/ruby/state_paths_not_writable_as_non_root.go [hunk 1]
```diff
@@ -0,0 +1,349 @@
...
+		return nil
+	}
+	loc := cmd.Location()
+	if len(loc) == 0 {
+		return nil
+	}
+	startLine := loc[0].Start.Line
+	// Position.Character is the 0-based column; rules.NewRangeLocation
+	// expects 1-based columns.
+	startCol := loc[0].Start.Character + 1
```
[the remark this answers is not in the document]
Comment (tinovyatkin) [reply internal/rules/tally/ruby/state_paths_not_writable_as_non_root.go]: Fixed in 976a8203 (the column off-by-one was the immediate cause of the broken snapshot). `rules.NewRangeLocation` columns are 0-based (per `internal/rules/location.go:60`), and `Position.Character` from buildkit's parser is also 0-based — so no conversion was needed. Snapshot regenerated; the fix now produces the correct `COPY --chown=rails:rails . .`.

File: internal/rules/tally/ruby/state_paths_not_writable_as_non_root.go [hunk 2]
```diff
@@ -0,0 +1,349 @@
...
+			}
+		}
+	}
+	for _, d := range railsStateDirs {
+		if !covered[d] {
+			return false
+		}
+	}
+	return true
+}
```
[the remark this answers is not in the document]
Comment (tinovyatkin) [reply internal/rules/tally/ruby/state_paths_not_writable_as_non_root.go:342]: Fixed in 976a8203. `chownCoversAllStateDirs` now also accepts `.` (current directory), `./`, `/rails`, and `/app` as covering all state dirs by construction — `chown -R rails .` and `chown -R rails /rails` recursively cover everything under WORKDIR, including the four canonical Rails state directories. The short-form `-R` and long-form `--recursive` were already both recognized by `chownIsRecursive`. Regression tests added: `chown -R rails . after COPY suppresses` and `chown -R rails:rails /rails after COPY suppresses`.

File: internal/rules/tally/ruby/state_paths_not_writable_as_non_root.go [hunk 3]
```diff
@@ -0,0 +1,349 @@
...
+		if src == "." || src == "./" || src == "/rails" || src == "/app" {
+			return true
+		}
+		// Deep paths into the project that imply directory content.
+		if strings.HasSuffix(src, "/.") {
+			return true
+		}
+	}
+	return false
+}
```
[the remark this answers is not in the document]
Comment (tinovyatkin) [reply internal/rules/tally/ruby/state_paths_not_writable_as_non_root.go:199]: Acknowledged but deferred. The current detection requires `COPY` source paths to be one of `.`, `./`, `/rails`, or `/app` — these are the patterns the corpus shows in 95%+ of Rails Dockerfiles. Adding `COPY my-app /app` (where `my-app` is a named directory in build context) is straightforward but expands the rule's surface area beyond what was reviewed in the corpus.

I've left this on the followup list. The current behavior is "skip the unusual case" rather than "false positive," which is the right direction given that this rule's recommendation (adding `--chown`) is non-trivial — over-eager triggering would be much worse than missing a niche pattern.

File: internal/rules/tally/ruby/state_paths_not_writable_as_non_root.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (tinovyatkin) [reply internal/rules/tally/ruby/state_paths_not_writable_as_non_root.go]: Fixed in 976a8203. Same root cause as the gemini high / greptile P1 — dropped the spurious `+ 1` on `loc[0].Start.Character`. `rules.NewRangeLocation` columns are 0-based (per `internal/rules/location.go:60`), so the parser's already-0-based column was correct. Snapshot regenerated; output is now valid Dockerfile syntax.

File: internal/rules/tally/ruby/state_paths_not_writable_as_non_root.go [hunk 4]
```diff
@@ -0,0 +1,349 @@
...
+		}
+		if !copyLooksLikeAppContent(copyCmd) {
+			continue
+		}
+		if copyHasChownToUser(copyCmd, user) {
+			continue
+		}
+		// If a later RUN performs `chown -R user dir(s)` covering all
+		// observed state dirs, the COPY is acceptable.
+		if stageHasRuntimeChown(sf, user) {
```
[the remark this answers is not in the document]
Comment (tinovyatkin) [reply internal/rules/tally/ruby/state_paths_not_writable_as_non_root.go]: Fixed in 976a8203. Renamed `stageHasRuntimeChown` to `stageHasRuntimeChownAfter`; it now takes the COPY's source line and only considers RUNs that come at-or-after that line. A chown BEFORE the COPY no longer suppresses the violation.

Two regression tests added: `chown -R before COPY does NOT suppress (COPY would re-introduce root-owned files)` and the symmetric `chown -R rails . after COPY suppresses`.

File: internal/rules/tally/ruby/state_paths_not_writable_as_non_root_test.go [hunk 5]
```diff
@@ -0,0 +1,164 @@
...
+		t.Fatal("expected a suggested fix")
+	}
+	if v.SuggestedFix.Safety != rules.FixSuggestion {
+		t.Errorf("Safety = %v, want FixSuggestion", v.SuggestedFix.Safety)
+	}
+	got := v.SuggestedFix.Edits[0].NewText
+	want := "--chown=rails:rails "
+	if !strings.Contains(got, want) {
+		t.Errorf("fix should contain %q; got: %q", want, got)
+	}
```
[the remark this answers is not in the document]
Comment (tinovyatkin) [reply internal/rules/tally/ruby/state_paths_not_writable_as_non_root_test.go]: Fixed in 976a8203. The test now uses `fix.ApplyFix` to apply the edits and verifies the resulting Dockerfile contains the expected `COPY --chown=rails:rails . .` (not just that the edit text contains the substring). Also added a `len(Edits) == 0` guard before indexing.

This also serves as a regression test for the column off-by-one bug — if the column math is wrong again, the assertion on the applied output will catch it.

File: internal/rules/tally/ruby/state_paths_not_writable_as_non_root.go [hunk 6]
```diff
@@ -0,0 +1,349 @@
...
+			continue
```
[the remark this answers is not in the document]
Comment (tinovyatkin) [reply internal/rules/tally/ruby/state_paths_not_writable_as_non_root.go]: Fixed in 976a8203. `stageHasRuntimeChownAfter` now takes the offending COPY's source line and only considers RUNs at-or-after that line. The Dockerfile pattern you describe (`RUN chown -R …` followed by `COPY . .`) now correctly produces a violation, since the COPY re-introduces root-owned files. Regression test added.

File: internal/rules/tally/ruby/state_paths_not_writable_as_non_root.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (tinovyatkin) [reply internal/rules/tally/ruby/state_paths_not_writable_as_non_root.go]: Fixed in 976a8203. Same root cause as gemini high / codex P1 / your other comment — column off-by-one. Dropped the `+ 1`; snapshot regenerated.

File: internal/rules/tally/ruby/state_paths_not_writable_as_non_root.go (hunk 6 above)
[the remark this answers is not in the document]
Comment (tinovyatkin) [reply internal/rules/tally/ruby/state_paths_not_writable_as_non_root.go]: Fixed in 976a8203 — same change addresses both the codex and your concern. `stageHasRuntimeChownAfter` only considers RUNs at-or-after the COPY's source line.

File: internal/integration/fixtures/fix/ruby-state-paths-not-writable-as-non-root/fixed_1.snap.Dockerfile [hunk 7]
```diff
@@ -0,0 +1,5 @@
+FROM ruby:3.3-slim
+WORKDIR /rails
+USER rails:rails
+COPY .--chown=rails:rails  .
```
[the remark this answers is not in the document]
Comment (tinovyatkin) [reply internal/integration/fixtures/fix/ruby-state-paths-not-writable-as-non-root/fixed_1.snap.Dockerfile]: Fixed in 976a8203. The off-by-one column bug (the underlying cause) is fixed; snapshot regenerated to capture the corrected output:

```
FROM ruby:3.3-slim
WORKDIR /rails
USER rails:rails
COPY --chown=rails:rails . .
CMD ["bin/rails", "server"]
```
